src/mngt/ocf_mngt_cache.c

```diff
@@ -1953,7 +1963,8 @@ struct ocf_pipeline_properties ocf_mngt_cache_stop_pipeline_properties = {
 		OCF_PL_STEP(ocf_mngt_cache_stop_check_dirty),
 		OCF_PL_STEP(ocf_mngt_cache_stop_remove_cores),
 		OCF_PL_STEP(ocf_mngt_cache_stop_unplug),
-		OCF_PL_STEP(ocf_mngt_cache_stop_deinit_metadata),
+		OCF_PL_STEP(ocf_mngt_cache_deinit_metadata),
+		OCF_PL_STEP(ocf_mngt_cache_deinit_cache_volume),
```

Comment:
> if adapter wants to do something in io context, it should do it before putting io
I agree. I am considering a situation where the adapter calls io->end() and then performs some finalization activities before calling ocf_io_put(). This pull request in in current version opens window for metadata being deinitialized before adapter calls io_put():

adapter_cb(ocf_io *io, int err)
{
    // OCF will not attempt to deinit metadata at this point as there are internal reference counters 
    // (e.g. metadata ref counter) incremented on behalf of this io
    io->end(err);
    // now metadata is potentially potentially deinitialized, management thread is waiting in ocf_volume_close() 
    // for volume reference counter to drop to 0
    do_something(io);  
    io->put();  // will potentially resume management thread ocf_volume_close()  
}

Reply:
I agree that reference counting on the volume seems superficial in the presence of multiple cache reference counters (like metadata). This is a consequence of our design, if I recall correctly the fact that we do not have the required request context to decrement OCF reference counters upon I/O deinitialization rather than its completion.

Reply:
Calling `ocf_io_put()` after metadata is deinitialized is harmless, as this function doesn't touch metadata anyway.

Calling `ocf_io_put()` on completion path is not a problem. The problem is that this `ocf_io_put()` called in the adapter code, not in the OCF code. All the `io->end()` calls from adapter end up in OCF internal callbacks anyway, so I see no good reason why we shouldn't put the io there.

Reply:
I didn't say anything about ocf_io_put() being harmful after deinitializing metadata or the put() call being impossible to be moved to inside OCF. 

Actually I think this would be great idea to remove ocf_io_put() from API and it would invalidate my objections to this patch. But the version under review is what it is - adapter now has a window to do something dangerous between io->end() and ocf_io_put() - and this is not the case on the baseline.